File: 卡码/代码随想录/Go/图论/bfs/LeetCode_0130.go
// 130. 被围绕的区域
package main

func solve(board [][]byte) {
	n, m := len(board), len(board[0])
	dir := [2][4]int{[4]int{0, 1, 0, -1}, [4]int{1, 0, -1, 0}}
	var bfs func(x, y int)
	bfs = func(x, y int) {
		board[x][y] = 'Y'
		q := make([]int, 0)
		q = append(q, x, y)
		for len(q) > 0 {
			x, y = q[0], q[1]
			q = q[2:]
			for i := 0; i < 4; i++ {
				x1 := x + dir[0][i]
				y1 := y + dir[1][i]
				if x1 < 0 || x1 > n-1 || y1 < 0 || y1 > m-1 {
					continue
				}
				if board[x1][y1] == 'O' {
					q = append(q, x1, y1)
					board[x1][y1] = 'Y'
				}
			}
		}

	}
	for i := 0; i < n; i++ {
		if board[i][0] == 'O' {
			bfs(i, 0)
		}
		if board[i][m-1] == 'O' {
			bfs(i, m-1)
		}
	}
	for j := 0; j < m; j++ {
		if board[0][j] == 'O' {
			bfs(0, j)
		}
		if board[n-1][j] == 'O' {
			bfs(n-1, j)
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if board[i][j] == 'O' {
				board[i][j] = 'X'
			}
			if board[i][j] == 'Y' {
				board[i][j] = 'O'
			}
		}
	}
}

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/图论/bfs/LeetCode_0130.java
package com.kurumi.leetCode.codeThoughts.图论.bfs;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 130. 被围绕的区域
 */
public class LeetCode_0130 {
    static class Solution {
        int[][] dir = new int[][] { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };

        public void bfs(char[][] board, int x, int y) {
            board[x][y] = 'Y';
            Deque<Integer> q = new ArrayDeque<>();
            q.offer(x);
            q.offer(y);
            while (!q.isEmpty()) {
                x = q.poll();
                y = q.poll();
                for (int i = 0; i < 4; i++) {
                    int nextX = x + dir[i][0];
                    int nextY = y + dir[i][1];
                    if (nextX < 0 || nextX > board.length - 1 || nextY < 0 || nextY > board[0].length - 1)
                        continue;
                    if (board[nextX][nextY] == 'O') {
                        q.offer(nextX);
                        q.offer(nextY);
                        board[nextX][nextY] = 'Y';
                    }
                }
            }
        }

        public void solve(char[][] board) {
            int n = board.length;
            int m = board[0].length;
            for (int i = 0; i < n; i++) {
                if (board[i][0] == 'O')
                    bfs(board, i, 0);
                if (board[i][m - 1] == 'O')
                    bfs(board, i, m - 1);
            }
            for (int j = 0; j < m; j++) {
                if (board[0][j] == 'O')
                    bfs(board, 0, j);
                if (board[n - 1][j] == 'O')
                    bfs(board, n - 1, j);
            }
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (board[i][j] == 'O')
                        board[i][j] = 'X';
                    if (board[i][j] == 'Y')
                        board[i][j] = 'O';
                }
            }
        }
    }
}
